### PR TITLE
Node Plop 0.13.0

### PR DESCRIPTION
**Pull request by crutchcorn**
This PR rewrites the entirety of `node-plop` to ESM in preparation for `plop@3`. This means the implications from this Gist apply:

https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c

This means that we only support the following versions of Node:

`^12.20.0 || ^14.13.1 || >=16.0.0`

A few APIs have changed outside of that:

- `plop.load` returns a promise that must be `await`ed
- `plop()` returns a promise that must be `await`ed
- `promptBypass()` returns a promise that must be `await`ed
- TS types have been made much more accurate, let us know what's broken

### CODE REVIEW

**Comment by Pike** [comment]
I've had a quick glance, looks nice. I stumbled upon the TS definition changes, so a couple of thoughts.

I was wondering if the usage of `ActionType` could be generic? AKA, if you add custom action types, you'd pass a

```
type MyActionType = MyActionConfig | ActionType
```

I haven't actually given that an honest try, with TS that could just be a rabbit hole. I also don't want to scope creep, it's just an idea I had while working on a prototype using plop in TS. That had both custom prompts and custom actions, so I casted away a lot.

And thanks for dropping the optional args in `CustomActionFunction`, that felt so weird.

**Comment by crutchcorn** [comment]
Thanks so much for the feedback @Pike! Really appreciate it - It can be hard to get any thoughts of an upcoming major release of an OSS project :)

I guess my question is - is there a reason we couldn't do the `ActionType | CustomAction`? That's not using a generic input, but it does work.

I'm not sure where it would be used that a union `|` wouldn't work. Could you possibly give an example of what you're looking for? That way, I could add to our typing tests with an example?

File: tsconfig.json [hunk 1]
```diff
@@ -1,20 +1,9 @@
 // Used by eslint for linting
 {
+	"extends": "./types/tsconfig.json"
```
Comment (Pike) [inline tsconfig.json:3]: I guess this should have a trailing comma?

I've checked out the branch and tried to do stuff in vscode, and this file as well as the types stuff has errors I don't understand. Something seems to throw off vscode so early that it just says ", expected" (comma) as first thing in the file.

**Comment by Pike** [comment]
I've done a few local thought experiments, and I think that `CustomActionConfig` is already a great enhancement.

One thing I noticed, though, is that with the following patch, the test still pass:

```diff
diff --git a/types/test.ts b/types/test.ts
index 0c5851d..70394c3 100644
--- a/types/test.ts
+++ b/types/test.ts
@@ -221,7 +221,6 @@ function e(plop: NodePlopAPI) {
                prompts: [],
                actions: [{
                        type: 'doTheThing',
-                       configProp: 'available from the config param'
                } as CustomActionConfig<'doTheThing'>,
                {
                        type: 'doTheAsyncThing',
```

Now, I was successful to do

```ts
interface TheThingConfig extends CustomActionConfig<'doTheThing'> {
    configProp: string
}
```

and that use that in the cast in `actions`. But I wasn't successful to close the loop on that actually meaning something to the action function. I did a little bit of toying to make `setActionType` generic, and was successful in making the type generic, but not the config type. Aka, yes, in `setActionType<K extends string>()`, it would know that type was supposed to be `doTheThing`, but `configProp` was still totally disconnected.

Maybe that's a hint for you? I was already baffled with `Omit` :-)

I've also taken another look at my prototype, which is a bit of weird kid, as somewhere in the middle I used dynamic imports, which of course goes through `any`, and makes all things good. Looking back, it might be that the mapping between prompts and Answers was actually more tricky still. There's a bit of mismatch in `runPrompts` returning `any` and `runActions` taking `inquirer.Answers`? Maybe that's a thing to pull up to a generic on `PlopGenerator` that falls back to `inquirer.Answers`?

**Comment by crutchcorn** [comment]
@Pike I admittedly think that at some point we're exceeding the scope I have planned for `plop@3`. Adding plugin support for typing (which is basically what we're doing) is tough. Like, huge huge changes tough.

Let's move this convo to a new issue (in the `plop` repo, please. I'll be archiving this repo). If you could do that and link in this PR, that would be hugely helpful! And thank you for the continued thoughts/feedback!

File: tsconfig.json (hunk 1 above)
Comment (crutchcorn) [reply tsconfig.json:3]: I tend to use Webstorm, which doesn't have issues. But I'll check with VSC

**Comment by Pike** [comment]
Sure, sounds like a good call. I'll do that once 3 is out, just to not talk about a somewhat moving target.

Thanks for your work.

**Pull request merged**